### ISSUE TITLE
LibreOffice/OO: No cross-paragraph checks

### ISSUE THREAD

**Issue by danielnaber** [languagetool-org/languagetool#194]
[report copied from https://sourceforge.net/p/languagetool/bugs/40/] 

Rules like WordCoherencyRule don't seem to work inside LO/OO, probably because this code calls `getAllRules()`, which then resets the status of each rule:

```
paragraphMatches = langTool.check(paraText, false, JLanguageTool.paragraphHandling.ONLYPARA);
```

**Comment by danielnaber** [comment]
This bug still exists, but the explanation with `getAllRules()` resetting the rule status seems outdated.

**Comment by FredKruse** [comment]
This problem is solved as far as possible. The standard is a check of 5 paragraphs before and 5 paragraphs behind the one to check. A full text check isn't realistic because it cost minutes for a 600 sites document and has be done again after every change.
I think the issue could be closed.

**Comment by danielnaber** [comment]
@FredKruse Isn't this solved now? I mean, not just for a few paragraphs but for all text (because we have the mode that runs only text-level checks and that is fast)?

**Comment by FredKruse** [comment]
I did a modification on the text level rules, that I should explain:
The full text check for all text level rules is not so fast as I hoped. Now I added an abstract method for text level rules, that gives back the minimum paragraphs that had to be checked to give a correct result for the specific rule (n==-1 full text check is needed, n==0 only the current paragraph needs to be checked, n > 0 n paragraphs before and behind the current paragraph has to be checked.
Examples: 
n == -1 AbstractWordCoherencyRule, GenericUnpairedBracketsRule, ...
n == 0   MultipleWhitespaceRule, LongParagraphRule, ...
n >= 1   ParagraphRepeatBeginningRule (n == 1), WordRepeatBeginningRule (n == 2), ...
The LO extension checks only for the rules n < 0 the full text (this are only 4 rules). This is rather fast.
The other text level rules are checked  for n paragraphs.
This is needed to because there are three criteria for LO extension:
1. The over all performance (time period to check the whole text)
2. The time period till the first marks are shown (this is a very critical criteria. It is depends on the length of the text and the performance of the used CPU. For log text or older CPUs it takes to long to check the full text with all rules)
3. The time period till changes of texts disappear (here the same problems arise as discussed under 2.)
The work done is a compromise between the 3 points.
